framework/source/class/qx/core/Id.js

```diff
@@ -0,0 +1,166 @@
+     * with a call to `qx.core.Id.getObject()`.
+     * 
+     * This will return null if it is not possible to calculate a path because one of the
+     * ancestors has a null `objectId`.
+     * 
+     * This will also return null if the top-most ancestor is not one of the globals registered
+     * with `registerObject` or a known global (such as the application); however, by passing
+     * `false` as the `mustBeRegistered` parameter, this restriction is removed and a path will
+     * still be returned (it's just that in this case, you will have to know the the root object
+     * to query and `qx.core.Id.getObject` will not work on that path)  
+     * 
+     * @param obj {qx.core.Object} the object
+     * @param mustBeRegistered {Boolean} whether the
+     * @return {String} full path to the object
+     */
+    getAbsoluteIdOf: function(obj, mustBeRegistered) {
+      var segs = [];
+      var application = qx.core.Init.getApplication();
+      while (obj) {
+        var id = obj.getObjectId();
+        if (!id) {
+          this.error("Cannot determine an absolute Object ID because one of the ancestor ObjectID's is null");
+          return null;
+        }
+        segs.unshift(id);
+        var owner = obj.getOwner();
+        if (owner) {
+          // Find the ID of the owner, *if* it is registered as a top level object
+          var ownerId = null;
+          if (owner === application) {
+            ownerId = "application";
+          } else {
+            ownerId = this.__registeredIdHashes && this.__registeredIdHashes[owner.toHashCode()] || null;
+          }
+
+          // When we have found the ID of a top level object, add it to the path and stop 
+          if (ownerId) {
+            segs.unshift(ownerId);
+            break;
+          }
+        } else {
+          if (mustBeRegistered === false) {
+            break;
+          }
+          this.error("Cannot determine a global absolute Object ID because the topmost object is not registered");
+          return null;
+        }
+        obj = owner;
+      }
+      var path = segs = segs.join("/");
+      return path;
+    },
+    
+    /**
+     * Registers an object with an ID; as this is registering a global object which is the root of a tree 
+     * of objects with IDs, the `id` parameter can be provided to set the ID used for the root object - this
+     * allows an object to be registered under a well known, common name without affecting the API of the
+     * object.  
+     * 
+     * @param obj {qx.core.Object} the object to register
+     * @param id {String?} the ID to register the object under, otherwise the object's own Object Id is used
+     */
+    register: function(obj, id) {
+      if (!this.__registeredObjects) {
+        this.__registeredObjects = {};
+        this.__registeredIdHashes = {};
+      }
+      if (!id) {
+        id = obj.getObjectId();
+      }
+      this.__registeredObjects[id] = obj;
```

Comment:
Why doesn't this create the same problem we had previously, with the registry, of preventing automatic garbage collection because the object reference is retained here?

Reply:
Because this is only a list of top level objects, not an exhaustive list of *all* objects.

EG the idea is that objects with public IDs are 'scoped" within their parent, so that the path "A/B/C/D" means that the object "A" is registered globally and it has a pointer to B which has a pointer to C etc - and this means that B, C, and D are not in a global list and hence do not trigger the memory leaks that the ObjectRegistry lead to.

It is true that in order to garbage collect, you need to remove A from the global list (or if you just wanted to remove C and D, you'd remove C from B) - but this benefits from a cascade effect.